src/Umbraco.Commerce.PaymentProviders.Mollie/MollieOneTimePaymentProvider.cs

```diff
@@ -348,31 +354,45 @@ private async Task<CallbackResult> ProcessRedirectCallbackAsync(PaymentProviderC
             };
 
             PropertyValue mollieOrderId = ctx.Order.Properties["mollieOrderId"];
+            _logger.Info($"ProcessRedirectCallbackAsync; mollieOrderId: '{mollieOrderId}'; ctx.Order.OrderNumber: '{ctx.Order.OrderNumber}'.");
             using (var mollieOrderClient = new OrderClient(ctx.Settings.TestMode ? ctx.Settings.TestApiKey : ctx.Settings.LiveApiKey))
             {
-                OrderResponse mollieOrder = await mollieOrderClient.GetOrderAsync(mollieOrderId, true).ConfigureAwait(false);
-
-                if (mollieOrder.Embedded.Payments.All(x => x.Status == MolliePaymentStatus.Canceled))
-                {
-                    result.HttpResponse.Headers.Location = new Uri(ctx.Urls.CancelUrl);
-                }
-                else
+                OrderResponse mollieOrder = await mollieOrderClient.GetOrderAsync(mollieOrderId, true);
+                if (mollieOrder != null)
                 {
-                    // If the order is pending, Mollie won't sent a webhook notification so
-                    // we check for this on the return URL and if the order is pending, finalize it
-                    // and set it's status to pending before progressing to the confirmation page
-                    if (mollieOrder.Status.Equals("pending", StringComparison.InvariantCultureIgnoreCase))
+                    if (mollieOrder.Embedded.Payments.Last()?.Status == MolliePaymentStatus.Failed)
+                    {
+                        // Mollie redirects user here when the payment failed
+                        result.HttpResponse.Headers.Location = new Uri(ctx.Urls.ErrorUrl);
```

Comment:
Looking at the mollie docs it looks like they send a failure reason https://docs.mollie.com/docs/handling-errors-with-mollie-components I think we should look to add this to the `ErrorUrl` as a query string parameter such that folks can display a suitable error message.

Reply:
if I add a query parameter to the `ctx.Urls.ErrorUrl`, will it be passed to the error page in payment provider setting? Won't the `ctx.Urls.ErrorUrl` will be consumed by umbraco commerce?

Reply:
Good question. So we could append query params to `ctx.Urls.ErrorUrl` but by default you are right, they wouldn't automatically pass through. But payment providers have a `GetErrorUrl(PaymentProviderContext context)` method which you can override to resolve the local error URL for that route to navigate to. In this method then you could check `context.Request` to check the query string on the `ctx.Urls.ErrorUrl` and if there is a `reason` param, then append it to the local URL returned from `GetErrorUrl`. Does that make sense?